ZAPD/ZScalar.cpp

```diff
@@ -0,0 +1,224 @@
+	default:
+		return "char*";
+	}
+}
+
+int ZScalar::MapTypeToSize(const ZScalarType scalarType)
+{
+	switch (scalarType) {
+	case ZSCALAR_S8:
+		return sizeof(scalarData.s8);
+	case ZSCALAR_U8:
+		return sizeof(scalarData.u8);
+	case ZSCALAR_S16:
+		return sizeof(scalarData.s16);
+	case ZSCALAR_U16:
+		return sizeof(scalarData.u16);
+	case ZSCALAR_S32:
+		return sizeof(scalarData.s32);
+	case ZSCALAR_U32:
+		return sizeof(scalarData.u32);
+	case ZSCALAR_S64:
+		return sizeof(scalarData.s64);
+	case ZSCALAR_U64:
+		return sizeof(scalarData.u64);
+	case ZSCALAR_F32:
+		return sizeof(scalarData.f32);
+	case ZSCALAR_F64:
+		return sizeof(scalarData.f64);
+	default:
+		return 0;
+	}
+}
+
+int ZScalar::GetRawDataSize()
+{
+	return ZScalar::MapTypeToSize(scalarType);
+}
+
+void ZScalar::ParseRawData()
+{
+	ZScalar::ParseRawData(rawData, rawDataIndex);
+}
+
+void ZScalar::ParseRawData(const std::vector<uint8_t>& data, const int offset)
+{
+	switch (scalarType) {
+	case ZSCALAR_S8:
+		scalarData.s8 = BitConverter::ToInt8BE(data, offset);
+	case ZSCALAR_U8:
+		scalarData.u8 = BitConverter::ToUInt8BE(data, offset);
+	case ZSCALAR_S16:
+		scalarData.s16 = BitConverter::ToInt16BE(data, offset);
+	case ZSCALAR_U16:
+		scalarData.u16 = BitConverter::ToUInt16BE(data, offset);
+	case ZSCALAR_S32:
+		scalarData.s32 = BitConverter::ToInt32BE(data, offset);
+	case ZSCALAR_U32:
+		scalarData.u32 = BitConverter::ToUInt32BE(data, offset);
+	case ZSCALAR_S64:
+		scalarData.s64 = BitConverter::ToInt64BE(data, offset);
+	case ZSCALAR_U64:
+		scalarData.u64 = BitConverter::ToUInt64BE(data, offset);
+	case ZSCALAR_F32:
+		scalarData.f32 = BitConverter::ToFloatBE(data, offset);
+	case ZSCALAR_F64:
+		scalarData.f64 = BitConverter::ToDoubleBE(data, offset);
+	}
+}
+
+std::string ZScalar::GetSourceTypeName()
+{
+	return ZScalar::MapScalarTypeToOutputType(scalarType);
+}
+
+std::string ZScalar::GetSourceValue()
+{
+	switch (scalarType) {
+	case ZSCALAR_S8:
+		return StringHelper::Sprintf("%hhd", scalarData.s8);
+	case ZSCALAR_U8:
+		return StringHelper::Sprintf("%hhu", scalarData.u8);
+	case ZSCALAR_S16:
+		return StringHelper::Sprintf("%hd", scalarData.s16);
+	case ZSCALAR_U16:
+		return StringHelper::Sprintf("%hu", scalarData.u16);
+	case ZSCALAR_S32:
+		return StringHelper::Sprintf("%ld", scalarData.s32);
+	case ZSCALAR_U32:
+		return StringHelper::Sprintf("%lu", scalarData.u32);
+	case ZSCALAR_S64:
+		return StringHelper::Sprintf("%lld", scalarData.s64);
+	case ZSCALAR_U64:
+		return StringHelper::Sprintf("%llu", scalarData.u64);
+	case ZSCALAR_F32:
+		return StringHelper::Sprintf("%f", scalarData.f32);
+	case ZSCALAR_F64:
+		return StringHelper::Sprintf("%lf", scalarData.f64);
+	default:
+		return "\"BADTYPE!!\"";
+	}
+}
+
+std::string ZScalar::GetSourceOutputCode(const std::string& prefix)
+{
+	if (parent != nullptr)
+	{
+		parent->declarations[rawDataIndex] = new Declaration(DeclarationAlignment::None, 16, GetSourceTypeName(), name, false, GetSourceValue());
```

Comment:
Rather than writing to parent->declarations[] directly, please use one of the functions, such as parent->AddDeclaration(...). Accessing the dictionary directly is being depricated.